src/goto-instrument/synthesizer/expr_enumerator.h

```diff
@@ -0,0 +1,511 @@
+/*******************************************************************\
+Module: Enumerator Interface
+Author: Qinheping Hu
+\*******************************************************************/
+
+/// \file
+/// Enumerator Interface
+
+#ifndef CPROVER_GOTO_INSTRUMENT_SYNTHESIZER_EXPR_ENUMERATOR_H
+#define CPROVER_GOTO_INSTRUMENT_SYNTHESIZER_EXPR_ENUMERATOR_H
+
+#include <util/std_expr.h>
+
+#include <langapi/language_util.h>
+
+#include "synthesizer_utils.h"
+
+#include <iostream>
+#include <map>
+
+typedef std::list<exprt> exprst;
+typedef std::list<unsigned int> partitiont;
+
+/// A base class for expression enumerators.
+class enumerator_baset
+{
+public:
+  explicit enumerator_baset(bool is_none_enumerator = false)
+    : is_none_enumerator(is_none_enumerator)
+  {
+  }
+
+  /// Set the size of expressions enumerated by this enumerator.
+  virtual void set_size(unsigned int new_size) = 0;
+
+  virtual exprst enumerate() = 0;
+  virtual enumerator_baset *clone() const = 0;
+
+  virtual ~enumerator_baset()
+  {
+  }
+
+  bool is_none_enumerator;
+
+protected:
+};
+
+typedef std::list<enumerator_baset *> enumeratorst;
+
+/// Enumerator that enumerates leaf expressions from a given list.
+/// Leaf expressions are complete expressions with no hole or placeholder
+class leaf_enumeratort : public enumerator_baset
+{
+public:
+  explicit leaf_enumeratort(const exprst &leaf_exprs, bool is_none = false)
+    : enumerator_baset(is_none), leaf_exprs(leaf_exprs)
+  {
+    expected_size = 0;
+  }
+
+  /// Enumerate expressions in the given list.
+  exprst enumerate() override
+  {
+    exprst result = exprst();
+
+    // Do nothing if the size of epxected enumerated expression is not 1.
+    if(expected_size != 1)
+      return result;
+
+    // Otherwise return expressions in the given list.
+    for(const auto &expr : leaf_exprs)
+    {
+      result.push_back(expr);
+    }
+    return result;
+  }
+
+  void set_size(unsigned int new_size) override
+  {
+    expected_size = new_size;
+  }
+
+  leaf_enumeratort *clone() const override
+  {
+    return new leaf_enumeratort(leaf_exprs, is_none_enumerator);
+  }
+
+  unsigned int get_expected_size()
+  {
+    return expected_size;
+  }
+
+protected:
+  /// Size of expected enumerated expressions. The size of a leaf expression
+  /// is 1. So we enumerate only when the `expected_size` is 1.
+  unsigned int expected_size;
+  exprst leaf_exprs;
```

Comment:
It seems this list is actually constant, only ever populated by the constructor? Is it even right for it to be a list of its own and not a reference? (Maybe yes, for the source object might go out of scope.)